### PR TITLE
tickets/DM-25421

### PR DESCRIPTION
**Pull request by mareuter**

### CODE REVIEW

File: bin/run_dsm.py
Comment (r-owen) [inline bin/run_dsm.py:17]: Please consider using `await salobj.set_summary_state(...)` to replace lines 15-18. It's a single call to go from any initial state to the desired state.

File: bin/run_dsm.py
Comment (r-owen) [inline bin/run_dsm.py:55]: I'm a bit worried about folks thinking -s means "simulate". It's a bit more typing, but you might consider `-s`/`--simulate` with a required value "fast" or "slow". Or pick one to be the default.

File: bin/shutdown_dsm.py
Comment (r-owen) [inline bin/shutdown_dsm.py:10]: Another good use of `set_summary_state`. It will work from any initial state, and OFFLINE works as expected.

File: doc/conf.py
Comment (r-owen) [inline doc/conf.py:14]: Please consider adding ts_salobj, as well.

File: doc/conf.py
Comment (r-owen) [inline doc/conf.py:10]: I strongly recommend using `lsst.ts.dsm.__version__` so you don't need the `version.py` file.

File: doc/developer-guide/developer-guide.rst
Comment (r-owen) [inline doc/developer-guide/developer-guide.rst:44]: Please add the link (or remove this note)

File: doc/developer-guide/developer-guide.rst
Comment (r-owen) [inline doc/developer-guide/developer-guide.rst:52]: I find this recommendation rather confusing. I think you are saying that you should clone the repo into a directory that you use for other repositories. Please either reword or consider deleting it (this is standard git practice).

My personal recommendation is to clone into a directory that you can mount in your Docker image. But that's all explained in the docs on how to use these Docker development images. Best not to go into that here.

File: doc/developer-guide/developer-guide.rst
Comment (r-owen) [inline doc/developer-guide/developer-guide.rst:54]: This tag and all those notes about SAL 5, etc. will go out of date too quickly. Please don't do that. It's just asking for the docs to be frustratingly out of date.

If `:develop` will work, you can recommend that. Otherwise I suggest a link to somewhere that shows what versions are available and leaving it at that.

I suggest documenting requirements somewhere else (e.g. the version history or a standard requirements file, if our build system supports that).

File: doc/developer-guide/developer-guide.rst
Comment (r-owen) [inline doc/developer-guide/developer-guide.rst:82]: session -> section

File: bin/shutdown_dsm.py
Comment (r-owen) [reply bin/shutdown_dsm.py:10]: That said...I'm curious why you have a shutdown script? I have never seen that done before.

Perhaps we should have a generic one in salobj?

File: doc/index.rst
Comment (r-owen) [inline doc/index.rst:6]: Please remove all boilerplate notes. It makes the real content much harder to read.

File: doc/index.rst
Comment (r-owen) [inline doc/index.rst:82]: Your User Guide is reasonably short; please consider making it part of this page instead of a separate page. That makes the most used information most readily available. This is especially useful because the links in the navigation bar along the left edge can't jump to other pages (which is really frustrating).

See https://ts-atdome.lsst.io for an example.

File: doc/user-guide/user-guide.rst
Comment (r-owen) [inline doc/user-guide/user-guide.rst:13]: Again please delete all the boilerplate

File: doc/developer-guide/developer-guide.rst
Comment (r-owen) [inline doc/developer-guide/developer-guide.rst:86]: I strongly suggest that "how to run the DSM", "how to run it in simulation mode" and "how to configure" be part of the User Guide. 

My mental model is:
* User: somebody interacts with, runs or stops the CSC.
* Developer: somebody who modifies the code.

File: python/lsst/ts/dsm/dsm_csc.py
Comment (r-owen) [inline python/lsst/ts/dsm/dsm_csc.py:95]: Please consider:
```
self.telemetry_directory = os.environ.get("DSM_TELEMETRY_DIR", "/home/saluser/telemetry")
```
Or, better yet, make the default value a module constant defined near the top of the file,
where you can easily see it and document it:
```
# Default telemetry directory if running in real mode and $DSM_TELEMETRY_DIR is not defined
DEFAULT_TELEMETRY_DIR = "/home/saluser/telemetry"
```

File: python/lsst/ts/dsm/dsm_csc.py
Comment (r-owen) [inline python/lsst/ts/dsm/dsm_csc.py:196]: Out of scope, but please consider (eventually) making this an event instead of a telemetry topic. Surely it is does not need to be output at frequent intervals? If it can change over time, but only sporadically, then an event is still a good choice. You can poll it and set the values as often as you like, and the event will only be output if any field changes.

File: python/lsst/ts/dsm/utils.py
Comment (r-owen) [inline python/lsst/ts/dsm/utils.py:25]: In the doc string please document what this converts in_time **to**. What does it return?

Interesting note below. How does that work? Once the line above has run I would expect ptime to be in UTC and that's the end of it -- especially if in_time can only be a float and not an astropy.time.Time.

File: python/lsst/ts/dsm/utils.py
Comment (r-owen) [inline python/lsst/ts/dsm/utils.py:52]: What you have works, but I gently suggest that dict's keyword constructor would make it a bit easier to read:
```
content = dict(
    timestamp = Time.now().isot,
    ui_versions = dict(
        code = ui_version,
        config = ui_config_version,
        config_file = ui_config_file,
    ),
    camera = dict(name=camera_name, fps=camera_fps),
    data = dict(buffer_size=data_buffer_size, acquisition_time=data_acquisition_time),
)
```

File: setup.cfg
Comment (r-owen) [inline setup.cfg:7]: Why exclude doc? It has at least one .py file (conf.py).

Also please consider `__init__.py` and `version.py` without the prefixes.

File: ups/ts_dsm.table
Comment (r-owen) [inline ups/ts_dsm.table:7]: Please make a separate pull request to remove DSM from ts_config_eas.

**Review by r-owen — APPROVED** [review]
This looks great.

I have some changes I hope you will consider: especially using salobj.set_summary_state and consider making how to stop/start the CSC part of the User's Guide and make it part of the main documentation page.

In one of the doc pages you say that the CSC can be configured, but I am not sure what you were referring to. The env variable?

File: bin/run_dsm.py
Comment (mareuter) [reply bin/run_dsm.py:17]: Does the call transition through all intervening states? I can't remember if it does.

File: ups/ts_dsm.table
Comment (mareuter) [reply ups/ts_dsm.table:7]: Thanks for the reminder. I did the work earlier. You should have received a PR for review.

File: bin/run_dsm.py
Comment (r-owen) [reply bin/run_dsm.py:17]: Yes.

File: ups/ts_dsm.table
Comment (r-owen) [reply ups/ts_dsm.table:7]: Great! I just reviewed it.

File: setup.cfg
Comment (mareuter) [reply setup.cfg:7]: I'm a bit confused about the "without the prefixes" suggestion. I took off the *.py from those in the config and it didn't seem to have any effect.

File: bin/run_dsm.py
Comment (mareuter) [reply bin/run_dsm.py:55]: Since I don't use the amain form of executor since it does not play well with the auto transition function, the --simulate flag doesn't show up if you run --help on the scripts.

File: bin/shutdown_dsm.py
Comment (mareuter) [reply bin/shutdown_dsm.py:10]: I have a shutdown script because the CSC is run locally on a laptop that is connected to hardware. I don't want to have to access LOVE or a Nublado instance to turn it off.

File: doc/developer-guide/developer-guide.rst
Comment (mareuter) [reply doc/developer-guide/developer-guide.rst:86]: This type of suggestion needs to be run back up to the documentation effort. It's a fundamental shift in how the docs are organized and the boilerplate structure does not reflect this.

File: bin/run_dsm.py
Comment (mareuter) [reply bin/run_dsm.py]: Done.

File: bin/shutdown_dsm.py
Comment (mareuter) [reply bin/shutdown_dsm.py]: Changed to set_summary_state.

File: doc/conf.py
Comment (mareuter) [reply doc/conf.py:10]: Done.

File: doc/conf.py
Comment (mareuter) [reply doc/conf.py]: Done.

File: python/lsst/ts/dsm/dsm_csc.py
Comment (mareuter) [reply python/lsst/ts/dsm/dsm_csc.py:196]: OK. I'll put it on the list to do on the next iteration since the next XML release is due out soon.

File: python/lsst/ts/dsm/utils.py
Comment (mareuter) [reply python/lsst/ts/dsm/utils.py:25]: The return value is documented as a float. The in_time parameter is documented as a string. The issue comes trying to derive a float from the astropy.time object. There is no way to get a timestamp (float) without going through the datetime intermediate object. This has the issue that the assumed time zone in the datetime object is the local time zone where the conversion happens. This means it is inconsistent in the timestamp return without doing the replace action as the timestamp will change depending on where the machine doing the conversion is located.

File: python/lsst/ts/dsm/utils.py
Comment (mareuter) [reply python/lsst/ts/dsm/utils.py]: Done.

File: doc/developer-guide/developer-guide.rst
Comment (r-owen) [reply doc/developer-guide/developer-guide.rst:76]: For what it's worth: I already received permission to do it that way for my docs and am writing all my CSC docs that way. So a good fraction of our CSCs will be documented that way.

File: doc/index.rst
Comment (r-owen) [reply doc/index.rst:41]: Note that this is explicitly allowed in the doc design. The template assumes a separate page, but that is no barrier to doing it the other way.

File: python/lsst/ts/dsm/utils.py
Comment (r-owen) [reply python/lsst/ts/dsm/utils.py:25]: Thanks. I missed the doc string for in_time because github didn't show it (nor the brief description) and I forgot to expand the text.

Please expand on what is returned. I think it is TAI unix seconds? If so, consider using salobj's `tai_from_utc` function:
```
return salobj.tai_from_utc(ptime)
```
For what it's worth, astropy 4.2 will have built-in support for `unix_tai`, so eventually you can write this as `return ptime.unix_tai`. (But unless they get the speed up salobj will continue to perform its own computation, as slowness in astropy's handling of TAI was a critical problem for salobj performance).

File: python/lsst/ts/dsm/utils.py
Comment (r-owen) [inline python/lsst/ts/dsm/utils.py:38]: Description?

File: setup.cfg
Comment (r-owen) [reply setup.cfg:7]: Concretely I suggest changing `**/*/__init__.py` to `__init__.py` and `**/*/version.py` to `version.py`.

File: setup.cfg
Comment (r-owen) [inline setup.cfg:17]: Please consider allowing flake8 to run on this file. Append `  # noqa` to each intersphinx mapping line to eliminate the complaints about those. Everything else should be compliant.

File: bin/run_dsm.py
Comment (r-owen) [reply bin/run_dsm.py:51]: I am suggesting changing your API to match the standard CSC API: --simulate with a value. I agree that you will have to implement it yourself, but it's a minor change to your existing code.

**Review by r-owen — APPROVED** [review]
Thank you for making the updates. I have a few more suggestions I hope you will consider, but do as you like with them.

File: bin/run_dsm.py
Comment (mareuter) [reply bin/run_dsm.py:42]: Done.

File: setup.cfg
Comment (mareuter) [reply setup.cfg]: Done.

File: python/lsst/ts/dsm/utils.py
Comment (mareuter) [reply python/lsst/ts/dsm/utils.py]: Done.

File: python/lsst/ts/dsm/dsm_csc.py
Comment (mareuter) [reply python/lsst/ts/dsm/dsm_csc.py:95]: Done.

File: python/lsst/ts/dsm/utils.py
Comment (mareuter) [reply python/lsst/ts/dsm/utils.py:25]: I can check that out when the new astropy becomes available. In the meantime, using salobj.tai_from_utc(ptime) requires all of the microseconds in my unit test to be modified in order to pass. That seems a bit weird. Can you comment on why that would be the case?

Comment (r-owen) [reply python/lsst/ts/dsm/utils.py:25]: I don't know. But microseconds is a small number when time is represented as a float, so it's not entirely surprising. I'd be tempted to test for approximate equality.

BTW: `unix_tai` is actually new AstroPy 6.1, not 6.2. https://docs.astropy.org/en/stable/changelog.html

Comment (mareuter) [reply python/lsst/ts/dsm/utils.py:26]: Done.

**Comment by mareuter** [comment]
I think I have addressed all the comments. I want to leave the telemetry to event switch over for another PR. This one is already getting hard to follow. That will happen soon (on another PR) since XML 7.0 is coming soon.

**Review by r-owen — APPROVED** [review]
Looks great! I have a minor suggestion or two to consider before merging; do with them as you like.

**Comment by r-owen** [comment]
You might consider configuring this repo so that minor changes do not invalidate an existing approving review.
In Branches configuration uncheck this box:

`Dismiss stale pull request approvals when new commits are pushed`

**Comment by mareuter** [comment]
Thanks for the tip. I've gone an unchecked that option.

**Pull request merged**
